### PR TITLE
Merge upstream

### LINKED ISSUE #298: Backport Josh's correctness fixes
Lots of correctness stuff coming in over the past few days: https://issues.apache.org/jira/issues/?jql=project%20%3D%20SPARK%20AND%20reporter%20in%20(joshrosen)%20ORDER%20BY%20key%20DESC%2C%20priority%20DESC%2C%20updated%20DESC

- https://issues.apache.org/jira/browse/SPARK-22982
- https://issues.apache.org/jira/browse/SPARK-22983
- https://issues.apache.org/jira/browse/SPARK-22984
- https://issues.apache.org/jira/browse/SPARK-22985
- https://issues.apache.org/jira/browse/SPARK-22997

### PR DESCRIPTION
**Pull request by robert3005**
Fixes #298

To get all the 2.3.0 release goodies and correctness

### CODE REVIEW

**Comment by robert3005** [comment]
@ash211 @mccheah when you have time. Last commit has conflict resolution. I essentially reverted to our k8s resource manager code and only took some cleanup code in sparksubmit

**Comment by ash211** [comment]
Circle's failing:

```
org.apache.maven.project.ProjectBuildingException: Some problems were encountered while processing the POMs:
[WARNING] 'dependencyManagement.dependencies.dependency.(groupId:artifactId:type:classifier)' must be unique: org.bouncycastle:bcprov-jdk15on:jar -> version ${bouncycastle.version} vs 1.58 @ org.apache.spark:spark-parent_2.11:2.4.0-SNAPSHOT, /home/ubuntu/spark/pom.xml, line 1199, column 19
[FATAL] Non-resolvable parent POM for org.apache.spark:spark-dist_2.11-hadoop-palantir:[unknown-version]: Could not find artifact org.apache.spark:spark-parent_2.11:pom:2.3.0-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 21, column 11
[WARNING] 'dependencyManagement.dependencies.dependency.(groupId:artifactId:type:classifier)' must be unique: org.bouncycastle:bcprov-jdk15on:jar -> version ${bouncycastle.version} vs 1.58 @ org.apache.spark:spark-parent_2.11:2.4.0-SNAPSHOT, /home/ubuntu/spark/pom.xml, line 1199, column 19
[WARNING] 'dependencyManagement.dependencies.dependency.(groupId:artifactId:type:classifier)' must be unique: org.bouncycastle:bcprov-jdk15on:jar -> version ${bouncycastle.version} vs 1.58 @ line 1199, column 19
```

I think the bouncycastle dep here is a duplicate:

https://github.com/palantir/spark/pull/299/files#diff-600376dffeb79835ede4a0b285078036R1199

**Comment by robert3005** [comment]
Ah, haven't bumped versions everywhere. Will fixup on Monday

**Comment by onursatici** [comment]
Seems like `KubernetesDriverSpec` and `Config ` we have under `org.apache.spark.deploy.k8s.submit` shadows imports with the same name from k8s.

We also now have a class called `KubernetesClientApplication` from upstream which is basically a duplicate of our `k8s/Client`.

@mccheah for SA on conflict resolution

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 1]
```diff
@@ -93,6 +93,14 @@ object SparkSubmit extends CommandLineUtils with Logging {
 
   private val CLASS_NOT_FOUND_EXIT_STATUS = 101
 
+  // Following constants are visible for testing.
+  private[deploy] val YARN_CLUSTER_SUBMIT_CLASS =
+    "org.apache.spark.deploy.yarn.YarnClusterApplication"
+  private[deploy] val REST_CLUSTER_SUBMIT_CLASS = classOf[RestSubmissionClientApp].getName()
+  private[deploy] val STANDALONE_CLUSTER_SUBMIT_CLASS = classOf[ClientApp].getName()
+  private[deploy] val KUBERNETES_CLUSTER_SUBMIT_CLASS =
+    "org.apache.spark.deploy.k8s.submit.KubernetesClientApplication"
```
Comment (mccheah) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Use the old code here - `org.apache.spark.deploy.k8s.Client` I think.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala [hunk 2]
```diff
@@ -0,0 +1,238 @@
+/*
```
Comment (mccheah) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala]: Remove this.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkPodInitContainer.scala [hunk 3]
```diff
@@ -0,0 +1,116 @@
+/*
```
Comment (mccheah) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkPodInitContainer.scala]: We renamed all the packages to `k8s` recently and I think this is the result of that...

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,52 @@
+#
```
Comment (mccheah) [inline resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: We can't take the Dockerfile changes yet.

**Comment by mccheah** [comment]
I took a closer look and think the Kubernetes integration will actually break for us.

What we should do is `git rebase -i` and strip out all of the Kubernetes related commits that went into apache/master. Think that would give us a more accurate representation of what we want rather than trying to resolve conflicts manually.

**Comment by ash211** [comment]
I think we don't want to pull in ANY changes from apache/master in the `resources-managers/kubernetes` folder.  Then at some point in the future we drop everything in that folder and replace it with the version in apache/master.

Not with a rebase though (that breaks commit hashes so you can't tell if a commit is in a Palantir release or not) but with another replacement commit

**Comment by ash211** [comment]
PR incoming to this branch to show what I mean

**Comment by mccheah** [comment]
@aash think we want to be safe and avoid changes to `SparkSubmit` also.

**Comment by ash211** [comment]
See https://github.com/palantir/spark/pull/300 -- haven't looked at the `SparkSubmit` changes.  What in there should we be concerned about?

**Comment by mccheah** [comment]
https://github.com/palantir/spark/pull/299#pullrequestreview-88921715

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 1 above)
Comment (ash211) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: looks like it's `org.apache.spark.deploy.k8s.submit.Client` -- making that change on my PR

**Comment by ash211** [comment]
@robert3005 please do https://github.com/palantir/spark/pull/300/commits/b22c7764b8c0dbeec77685d6e2e2142ea6005f5c also

**Comment by ash211** [comment]
fysa @robert3005 pushed a commit to capture the failed mvn-install log

**Comment by robert3005** [comment]
@mccheah @ash211 @onursatici are these valid?
```
[error]  * method productElement(Int)java.lang.Object in object org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages#RetrieveSparkAppConfig does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages#RetrieveSparkAppConfig.productElement")
[error]  * method productArity()Int in object org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages#RetrieveSparkAppConfig does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages#RetrieveSparkAppConfig.productArity")
[error]  * method canEqual(java.lang.Object)Boolean in object org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages#RetrieveSparkAppConfig does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages#RetrieveSparkAppConfig.canEqual")
[error]  * method productIterator()scala.collection.Iterator in object org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages#RetrieveSparkAppConfig does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages#RetrieveSparkAppConfig.productIterator")
[error]  * method productPrefix()java.lang.String in object org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages#RetrieveSparkAppConfig does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages#RetrieveSparkAppConfig.productPrefix")
[error]  * method toString()java.lang.String in object org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages#RetrieveSparkAppConfig is declared final in current version
[error]    filter with: ProblemFilters.exclude[FinalMethodProblem]("org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages#RetrieveSparkAppConfig.toString")
```

**Comment by ash211** [comment]
It is correctly identifying a change in API of that class which was introduced in  https://github.com/apache-spark-on-k8s/spark/pull/224/files#diff-ec2c1511cb40d6f7310e819e2b53b345R31 and brought into Palantir Spark.

It's not in Apache yet: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedClusterMessage.scala#L31

This landed back in July though -- https://github.com/palantir/spark/commit/8ea3cf0 -- curious that MIMA is only just now flagging it.  Maybe they started checking these classes for backcompat recently?

**Comment by robert3005** [comment]
Maybe it's because we tagged a new version? I imagine we should add these to excludes

**Comment by ash211** [comment]
Yeah I think we should just exclude this class from mima checks -- the change was on the in-cluster message objects, which shouldn't actually have compatibility requirements since the versions are always in sync between driver and executors of a Spark cluster.

I think that's an addition to `project/MimaExcludes.scala` ?

**Comment by robert3005** [comment]
[WARN] Unable to detect inner functions for class:org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages.RetrieveSparkAppConfig is the issue - now need to figure out why generating mimaignores for this class fails

**Comment by ash211** [comment]
I don't see those ignores in this branch -- where did `org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages#RetrieveSparkAppConfig#` come from?

**Comment by robert3005** [comment]
There's GenerateMIMAIgnore that produces list of package private things to ignore. This class should be covered by this list but somehow is treated differently when scanning classpath. I will poke over weekend.

**Comment by robert3005** [comment]
So the fail is legit in that in 2.2.0 this object is a case object and it got converted to case class. Excludes are generated from previous version and not current so it's not ignored (as potentially new thing in api) - it's internal though so we should ignore. I am curious though why it wasn't flagging before - maybe previous version didn't get updated for mima checks?

**Comment by ash211** [comment]
Most recent errors look like some of our Parquet changes got reverted, or new tests got added that assume older parquet / bad timestamp format:

SPARK-22673: InMemoryRelation should utilize existing stats of the plan to be cached - org.apache.spark.sql.execution.columnar.InMemoryColumnarQuerySuite
moresbt.ForkMain$ForkError: org.scalatest.exceptions.TestFailedException: 700 did not equal 740
parquet timestamp conversion - org.apache.spark.sql.execution.datasources.parquet.ParquetInteroperabilitySuite
moresbt.ForkMain$ForkError: org.scalatest.exceptions.TestFailedException: int96TimestampConversion = false; vectorized = false INT64 did not equal INT96

**Comment by ash211** [comment]
Looked through these commits, which I think are the diff you made on top of the merge:

https://github.com/palantir/spark/pull/299/commits/2b9fdfe7c5fdfe42f7fdc7141e5317fb214c4439 https://github.com/palantir/spark/pull/299/commits/1f2962c7cdccded9b87fa9983e07689c1f38f6e6 https://github.com/palantir/spark/pull/299/commits/91a6f1f767cc0937265370757e9fc9a8c7983d71 https://github.com/palantir/spark/pull/299/commits/b8380fdaa7b63d33c8cfd75e9e727698aac489c1 https://github.com/palantir/spark/pull/299/commits/277425d576e9457a9b8b68998b45296544aaa038 https://github.com/palantir/spark/pull/299/commits/d37f4eeaa7d1e90aa3a1e48a88b3c0c4f13bb9b4 https://github.com/palantir/spark/pull/299/commits/fab5e5a0b604b9853d9a7cbbbeace1195549c2e2 https://github.com/palantir/spark/pull/299/commits/bd4056d901a56bdaef65f91f05b33d90915d463e https://github.com/palantir/spark/pull/299/commits/2cf672d7c4ee99a415c896ef334eff0d6822c957 https://github.com/palantir/spark/pull/299/commits/ab93a5363c18836a04aef0b865431c0b1a26e063

and I think things are good.  Will merge and cut a release!

**Pull request merged**